### PR TITLE
x509_vfy.c: Fix a regression in find_issuer(); extend and re-organize some tests

### PR DESCRIPTION
**Pull request by DDvO**
This fixes a regression reported for v1.1.1 in #13739 for the case that the candidate issuer cert is identical to the target cert.

Further related commits do the following:
* Add checks in case `X509_NAME_hash()` fails due to unsupported SHA1.
* Make `PEM_X509_INFO_read_bio_ex()` conservative on the error queue
* TEST: move cert, key, and CSR loading aux functions to new `testutil/load.c`

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: providers/implementations/storemgmt/file_store.c [hunk 1]
```diff
@@ -485,6 +485,8 @@ static int file_set_ctx_params(void *loaderctx, const OSSL_PARAM params[])
         BIO_snprintf(ctx->_.dir.search_name, sizeof(ctx->_.dir.search_name),
                      "%08lx", hash);
         X509_NAME_free(x509_name);
+        if (hash == 0)
```
Comment (t8m) [inline providers/implementations/storemgmt/file_store.c:488]: This is wrong as there is a small 1 in 2^64 chance that the real name hash value will be actually 0.

Comment (t8m) [reply providers/implementations/storemgmt/file_store.c:488]: Perhaps add ERR_peek_error() != 0 condition?

File: crypto/x509/x509_vfy.c [hunk 2]
```diff
@@ -327,14 +328,11 @@ static X509 *find_issuer(X509_STORE_CTX *ctx, STACK_OF(X509) *sk, X509 *x)
...
-        /*
-         * Below check 'issuer != x' is an optimization and safety precaution:
-         * Candidate issuer cert cannot be the same as the subject cert 'x'.
-         */
-        if (issuer != x && ctx->check_issued(ctx, x, issuer)
+        if (ctx->check_issued(ctx, x, issuer)
             && (((x->ex_flags & EXFLAG_SI) != 0 && sk_X509_num(ctx->chain) == 1)
                 || !sk_X509_contains(ctx->chain, issuer))) {
-            rv = issuer;
+            if (rv == NULL)
```
Comment (t8m) [inline crypto/x509/x509_vfy.c:334]: Why this if?

File: providers/implementations/storemgmt/file_store.c (hunk 1 above)
Comment (DDvO) [reply providers/implementations/storemgmt/file_store.c]: > This is wrong as there is a small 1 in 2^64 chance that the real name hash value will be actually 0.

I had also thought about the risk of the hash being legitimately 0.
When looking at the implementation of `X509_NAME_hash()`, one can see that the chance is actually 1 in 2^32,
and the value 0 is actually returned on error as well, which is a bad design.

> Perhaps add ERR_peek_error() != 0 condition?

This would work only if the error queue was/made empty before, which is not guaranteed/conservative,
and in either case not elegant.
Since I believe that "(unsigned) long" is meanwhile guaranteed to be actually 64 bits wide,
I've resorted to letting this function (and consequently also `X509_NAME_hash_old()`) return `0x100000000L` in case of error.

File: crypto/x509/x509_vfy.c (hunk 2 above)
Comment (DDvO) [reply crypto/x509/x509_vfy.c:334]: This is a minor "drive-by" improvement also mentioned in the adapted comment internally documenting the function:
> Prefer the first one that is not expired, else take the **first** expired one.

Before, it preferred the **last** expired one, which would be unexpected and inconsistent with other (partly documented) lookup behavior.

File: providers/implementations/storemgmt/file_store.c (hunk 1 above)
Comment (kaduk) [reply providers/implementations/storemgmt/file_store.c]: I don't think you can safely assume that long is 64 bits.
See, for example, the commit message for 66ecfb5437b22664a4358e8de6da942727f5fb11

Comment (DDvO) [reply providers/implementations/storemgmt/file_store.c]: I see, but then `0x100000000` collapses to `0`, which AFAICS is still the best one can do 
without changing the type of the function (in some not backward compatible way).

File: crypto/x509/x509_vfy.c (hunk 2 above)
Comment (t8m) [reply crypto/x509/x509_vfy.c:334]: But below you are calling x509_check_cert_time on the rv and not on the issuer. This looks wrong.

Comment (DDvO) [reply crypto/x509/x509_vfy.c:334]: Oh, indeed! Before the change it did not matter, but now it does.
Fixed.

File: crypto/x509/x509_vfy.c [hunk 3]
```diff
@@ -333,7 +333,7 @@ static X509 *find_issuer(X509_STORE_CTX *ctx, STACK_OF(X509) *sk, X509 *x)
                 || !sk_X509_contains(ctx->chain, issuer))) {
             if (rv == NULL)
                 rv = issuer;
-            if (x509_check_cert_time(ctx, rv, -1))
+            if (x509_check_cert_time(ctx, issuer, -1))
```
Comment (t8m) [inline crypto/x509/x509_vfy.c:336]: I suppose you should also return issuer directly here instead of break. As the rv should be returned only in case all check_cert_time calls fail.

Comment (t8m) [reply crypto/x509/x509_vfy.c:336]: And one more comment (sorry for that): Let's suppose we have a CA trusted cert store to which we add certificates at the end but never remove the old expired ones. It would be actually better to return the latest cert rather than the first one so I am still not fully convinced this is a right change.

Comment (DDvO) [reply crypto/x509/x509_vfy.c]: > I suppose you should also return issuer directly here instead of break. 

Right - fixed.
I recall that I had already done so, but this must have go lost when I distilled this PR from a larger set of fixes.

Comment (DDvO) [reply crypto/x509/x509_vfy.c]: > Let's suppose we have a CA trusted cert store to which we add certificates at the end but never remove the old expired ones. It would be actually better to return the latest cert rather than the first one so I am still not fully convinced this is a right change.

Well, to judge this properly we need to keep in mind that the preference of first vs. last only applies to certs that are currently invalid (such as, expired). If there is any other matching one that is currently valid, this one will be returned as the best choice.
So otherwise the worst that could happen in the scenario you mentioned is that we return that first expired cert that is older than in the last expired one in the list, so in practice not much of a difference because the verification won't succeed in either case.
If I revert to the original preference of taking the last expired one, which I am open to do if you still believe this is better,
we should document this exception to the normal rule of (in doubt) returning the first cert matching the respective criteria.

File: crypto/x509/x509_vfy.c [hunk 4]
```diff
@@ -325,16 +328,13 @@ static X509 *find_issuer(X509_STORE_CTX *ctx, STACK_OF(X509) *sk, X509 *x)
...
+        if (ctx->check_issued(ctx, x, issuer)
             && (((x->ex_flags & EXFLAG_SI) != 0 && sk_X509_num(ctx->chain) == 1)
                 || !sk_X509_contains(ctx->chain, issuer))) {
-            rv = issuer;
-            if (x509_check_cert_time(ctx, rv, -1))
-                break;
+            if (x509_check_cert_time(ctx, issuer, -1))
+                return issuer;
+            if (rv == NULL)
+                rv = issuer;
```
Comment (t8m) [inline crypto/x509/x509_vfy.c:337]: Could you please separate this out to a different PR as it is unrelated to the regression fix?

File: crypto/x509/x509_cmp.c [hunk 5]
```diff
@@ -257,9 +257,9 @@ unsigned long X509_NAME_hash(const X509_NAME *x)
 
     /* Make sure X509_NAME structure contains valid cached encoding */
     i2d_X509_NAME(x, NULL);
-    if (!EVP_Digest(x->canon_enc, x->canon_enclen, md, NULL, EVP_sha1(),
-                    NULL))
-        return 0;
+    if (!EVP_Digest(x->canon_enc, x->canon_enclen, md, NULL, EVP_sha1(), NULL))
+        /* TODO(3.0) This fails if SHA1 is not supported, e.g., in test_ssl_old */
+        return (unsigned long)0x100000000L; /* may collapse to 0 */
```
Comment (t8m) [inline crypto/x509/x509_cmp.c:262]: I still think this is not a proper solution as except for LP64 architectures this will return 0 as error. Or to be more precise, it is not a problem here but all the callers checking for 0 as failure value. If you really want to check for the error condition the only proper solution IMO is to add another API call which would have a proper error reporting and deprecate this one.

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/x509/x509_vfy.c (hunk 4 above)
Comment (DDvO) [reply crypto/x509/x509_vfy.c]: All right, I've extracted this change to #13805.

File: crypto/x509/x509_cmp.c (hunk 5 above)
Comment (DDvO) [reply crypto/x509/x509_cmp.c]: So I've spent the effort to add a new function `X509_NAME_hash_ex()` in order to be able to check 
if it failed due to unsupported SHA1 (or any other error reason), 
to deprecate `X509_NAME_hash()`,
and to document `X509_NAME_hash_ex()`, `X509_NAME_hash()` as well as the strongly related functions `X509_subject_name_hash()` and `X509_issuer_name_hash()`.

Comment (t8m) [reply crypto/x509/x509_cmp.c]: Thank you @DDvO for your effort. I am sorry for being so demanding. I'll start the final review pass right away.

File: crypto/x509/x509_cmp.c [hunk 6]
```diff
@@ -250,16 +250,20 @@ int X509_NAME_cmp(const X509_NAME *a, const X509_NAME *b)
     return ret < 0 ? -1 : ret > 0;
 }
 
-unsigned long X509_NAME_hash(const X509_NAME *x)
+unsigned long X509_NAME_hash_ex(const X509_NAME *x, int *ok)
```
Comment (t8m) [inline crypto/x509/x509_cmp.c:253]: I again apologize for being so demanding, but when you're adding a new _ex function anyway it would be a good idea to add libctx and propq arguments to the function allowing fetching the digest properly.

I am also not sure if returning the error by means of output argument will not be frowned upon by other commiters. I am personally OK with that as otherwise it would complicate changing the callers of the legacy function.

Comment (DDvO) [reply crypto/x509/x509_cmp.c]: > I again apologize for being so demanding, but when you're adding a new _ex function anyway it would be a good idea to add libctx and propq arguments to the function allowing fetching the digest properly.

I like the idea of further improving `X509_NAME_hash_ex()` this way - done.

> I am also not sure if returning the error by means of output argument will not be frowned upon by other commiters. I am personally OK with that as otherwise it would complicate changing the callers of the legacy function.

I see little other choice than adding such an output parameter since with 32-bit `long` there is no room in the return value space that could be used to indicate failure.

File: apps/rehash.c [hunk 7]
```diff
@@ -291,10 +291,23 @@ static int do_file(const char *filename, const char *fullpath, enum Hash h)
         goto end;
     }
     if (name != NULL) {
-        if ((h == HASH_NEW) || (h == HASH_BOTH))
-            errs += add_entry(type, X509_NAME_hash(name), filename, digest, 1, ~0);
+        if ((h == HASH_NEW) || (h == HASH_BOTH)) {
```
Comment (FdaSilvaYY) [inline apps/rehash.c:294]: nit : useless inside parenthesis.

File: doc/man3/X509_get_subject_name.pod [hunk 8]
```diff
@@ -74,6 +105,8 @@ earlier versions.
 X509_CRL_get_issuer() is a function in OpenSSL 1.1.0. It was previously
 added in OpenSSL 1.0.0 as a macro.
 
+X509_NAME_hash() was turned into a macro and deprecated in OpenSSL 3.0.
+
 =head1 COPYRIGHT
 
 Copyright 2015-2020 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (FdaSilvaYY) [inline doc/man3/X509_get_subject_name.pod:112]: nit: 2021 ?

File: test/testutil/load.c [hunk 9]
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (FdaSilvaYY) [inline test/testutil/load.c:2]: nit: 2021 ?

File: apps/rehash.c (hunk 7 above)
Comment (DDvO) [reply apps/rehash.c]: Thx, removed them.

File: doc/man3/X509_get_subject_name.pod (hunk 8 above)
Comment (DDvO) [reply doc/man3/X509_get_subject_name.pod]: I suppose some script will take care of required copyright changes, but done anyway.

File: test/testutil/load.c (hunk 9 above)
Comment (DDvO) [reply test/testutil/load.c]: Done likewise.

File: crypto/x509/x509_cmp.c [hunk 10]
```diff
@@ -250,20 +250,27 @@ int X509_NAME_cmp(const X509_NAME *a, const X509_NAME *b)
...
-        return 0;
-
-    ret = (((unsigned long)md[0]) | ((unsigned long)md[1] << 8L) |
-           ((unsigned long)md[2] << 16L) | ((unsigned long)md[3] << 24L)
-        ) & 0xffffffffL;
+    if (ok != NULL)
+        *ok = 0;
+    if (sha1 != NULL
+        && EVP_Digest(x->canon_enc, x->canon_enclen, md, NULL, sha1, NULL)) {
+        /* TODO(3.0) This fails if SHA1 is not supported, e.g., in test_ssl_old */
```
Comment (t8m) [inline crypto/x509/x509_cmp.c:266]: Can this TODO be removed now? IMO yes.

File: doc/man3/X509_get_subject_name.pod [hunk 11]
```diff
@@ -25,16 +34,29 @@ subject names
 
 =head1 DESCRIPTION
 
+X509_NAME_hash_ex() returns the SHA1 hash value of name B<x> or 0 on failure,
```
Comment (t8m) [inline doc/man3/X509_get_subject_name.pod:37]: `I<x>`

I know this is inconsistent with the unmodified text. But you're using `I<libctx>` anyway. I am not forcing you to change all the wrong occurrences in the file.

File: doc/man3/X509_get_subject_name.pod [hunk 12]
```diff
@@ -25,16 +34,29 @@ subject names
...
+using any given library context I<libctx> and property query I<propq>.
+The B<ok> result argument may be NULL
```
Comment (t8m) [inline doc/man3/X509_get_subject_name.pod:39]: `I<ok>`

File: doc/man3/X509_get_subject_name.pod [hunk 13]
```diff
@@ -25,16 +34,29 @@ subject names
...
+or else is used to return 1 for success and 0 for failure.
+Failure may happen on malloc error or if no SHA1 implementation is available.
+
+X509_NAME_hash() returns the SHA1 hash value of name B<x> or 0 on failure,
```
Comment (t8m) [inline doc/man3/X509_get_subject_name.pod:43]: `I<x>`

File: doc/man3/X509_get_subject_name.pod [hunk 14]
```diff
@@ -25,16 +34,29 @@ subject names
...
-X509_get_issuer_name() and X509_set_issuer_name() are identical to
-X509_get_subject_name() and X509_set_subject_name() except the get and
-set the issuer name of B<x>.
+X509_subject_name_hash() returns the SHA1 hash value of the subject name of
+certificate B<x>.
+
+X509_get_issuer_name(), X509_set_issuer_name(), and X509_issuer_name_hash()
+are identical to
+X509_get_subject_name(), X509_set_subject_name(), and X509_subject_name_hash()
+except they relate to the issuer name of B<x>.
```
Comment (t8m) [inline doc/man3/X509_get_subject_name.pod:59]: `I<x>`

File: providers/implementations/storemgmt/file_store.c [hunk 15]
```diff
@@ -481,10 +482,12 @@ static int file_set_ctx_params(void *loaderctx, const OSSL_PARAM params[])
         if (!OSSL_PARAM_get_octet_string_ptr(p, (const void **)&der, &der_len)
             || (x509_name = d2i_X509_NAME(NULL, &der, der_len)) == NULL)
             return 0;
-        hash = X509_NAME_hash(x509_name);
+        hash = X509_NAME_hash_ex(x509_name, NULL, NULL, &ok);
```
Comment (t8m) [inline providers/implementations/storemgmt/file_store.c:485]: Please use `ossl_prov_ctx_get0_libctx(ctx->provctx)` for the libctx. Keep propq at NULL.

**Review by t8m — CHANGES_REQUESTED** [review]
Last few nits to resolve.

File: crypto/x509/x509_cmp.c (hunk 10 above)
Comment (DDvO) [reply crypto/x509/x509_cmp.c]: I think so, too, meanwhile - done.

File: doc/man3/X509_get_subject_name.pod (hunk 11 above)
Comment (DDvO) [reply doc/man3/X509_get_subject_name.pod]: Ah, right; I should better have made all of them consistently use `I<`...`>`; done now.

File: providers/implementations/storemgmt/file_store.c (hunk 15 above)
Comment (DDvO) [reply providers/implementations/storemgmt/file_store.c]: Good point, done.

**Comment by DDvO** [comment]
> Last few nits to resolve.

Thanks; done.

File: doc/man3/X509_get_subject_name.pod [hunk 16]
```diff
@@ -25,16 +34,29 @@ subject names
 
 =head1 DESCRIPTION
 
-X509_get_subject_name() returns the subject name of certificate B<x>. The
+X509_NAME_hash_ex() returns the SHA1 hash value of name I<x> or 0 on failure,
```
Comment (t8m) [inline doc/man3/X509_get_subject_name.pod:37]: One more nit - as the value is not a full SHA1 hash, it should be mentioned here (or maybe even better in RETURN VALUES below).  Something like - `returns the first four bytes of SHA1 hash value of name I<x>. The value is converted to unsigned long in little endian order`.

Comment (DDvO) [reply doc/man3/X509_get_subject_name.pod]: Indeed, good point! 
Done in the RETURN VALUES section, and abstracted the description for `X509_NAME_hash_ex()`, `X509_NAME_hash()`, and `X509_subject_name_hash` accordingly.

File: doc/man3/X509_get_subject_name.pod [hunk 17]
```diff
@@ -68,7 +68,9 @@ X509_get_subject_name(), X509_get_issuer_name(), X509_REQ_get_subject_name()
 and X509_CRL_get_issuer() return an B<X509_NAME> pointer.
 
 X509_subject_name_hash() and X509_issuer_name_hash()
-return the respective hash value or 0 on failure.
+return the first four bytes of the SHA1 hash value,
```
Comment (t8m) [inline doc/man3/X509_get_subject_name.pod:71]: Please add X509_NAME_hash_ex and X509_NAME_hash here as well.

Comment (DDvO) [reply doc/man3/X509_get_subject_name.pod:72]: Oops, done.
Also squashed the recent fixup commits.

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Merged - thanks @t8m, also for shepherding further improvements of `X509_NAME_hash()`.
